### ISSUE TITLE
What is the best way to preserve ordering in a partition-topic using librdkafka?

### ISSUE THREAD

**Issue by tomciaaa** [edenhill/librdkafka#623]
Hi Magnus, 

I'm trying to use the librdkafka C++ API in a way that maintains the message ordering within a single topic partition, and I was wondering if it was currently possible to achieve this with librdkafka and batching.

My understanding is that to have a synchronous producer, I should follow the approach outlined in https://github.com/edenhill/librdkafka/wiki/Sync-producer which requires sending messages one by one. Trying to do this in batches I encountered a few apparent issues:
1. I cannot seem to send a `ProduceRequest` packet with a `MessageSet` larger than 254 messages, because that is statically calculated as `(IOV_MAX/4)-1`. I am not familiar with the iov structures and was wondering if this limit could be increased?
   The reason I find splitting a single batch into multiple `MessageSet` objects concerning is that this opens up the possibility of only some of the batch being committed by the broker due to transient errors and retrying would potentially reorder messages within a large batch.
2. Is there any plan to incorporate logic for maintaining batch ordering, so that the 'application->kafka' thread stops producing, until the preceding batches inside the topic partition are successfully delivered? 
   This would be very useful, as otherwise I end up redoing queue management and batching logic myself as I have to surround the code for producing a single batch in `Handle::pause` and `Handle::resume` method calls as well as adding appropriate time-outs.

Thanks

**Comment by edenhill** [comment]
Hey,

1. The IOV based limit is a bit of a concern in this use case, yes. I will try to find another solution to this. 

2. There are plans and an old issue to track that (https://github.com/edenhill/librdkafka/issues/204). There hasnt been that much fuzz over that lately but new interesting will promote it on the backlog. I agree it would be a useful feature.

**Comment by tomciaaa** [comment]
Thanks for the reply.

Good to hear preserving order is on the cards. If you don't mind, could you give a bit more explanation of why the IOV limit is divided by four and is it a bad idea to put more than one message in a single IOV entry -- the message size doesn't seem to be limited by IOV, as I have tested sending messages 8-1024 bytes big and all were subject to the same limits?

**Comment by edenhill** [comment]
librdkafka uses IOV for messages to provide zero-copying, which is important for performance when having large messages (for small messages it is probably counter productive).
The /4 is because each message is four iovecs: Header(part1), Key, Header(part2), Payload, it could easily be reduced to two  iovecs by simply copying the key, but this doesnt really solve the underlying problem. So a solution that does not depend on IOV_MAX is preferable (and not that hard to implement).

**Comment by edenhill** [comment]
issue #1: batches limited to 254 messages, has been fixed on master.

**Comment by edenhill** [comment]
you could probably do \#2 using max.in.flight=1 and batch.num.messages=<youdecide>

**Comment by aaronhays** [comment]
I'm trying to do something similar to number 2, but I'm not seeing a guaranteed way to do it with existing library features. My understanding may be flawed - perhaps you can check my logic.

Even setting `max.in.flight=1` don't you potentially have a problem if a MessageSet fails? Imagine librdkafka has one MessageSet inflight and one waiting, librdkafka gets an error back from the broker for the inflight MessageSet and now nothing is inflight. At this point, it would send the pending MessageSet, which might succeed, leaving a "gap", correct?

You could mitigate this by setting `retries` to a high number and hoping that transient issues (e.g. no leader) are resolved within  the retry window, but still no guarantee, right?

Thanks!

**Comment by edenhill** [comment]
If the single outstanding MessageSet fails producing and there are remaining retries it will be inserted back on the transmission queue before any subsequently added MessageSets.
If the retries are exhausted the messages in the messageSet will fail (delivery reports will be triggered for each one).

So I think it should work.

**Comment by aaronhays** [comment]
I'm certainly not deeply familiar with the code, but as I understand it, when client code produces a message, it goes onto a queue for the topic/partition. Once the size, count, or time from the configuration elapse, a buffer is prepared and placed onto a broker send queue. Then, provided that the size of a wait response queue is less than `max.inflights`, the next buffer on the send queue will be sent and an entry placed on the wait response queue (assuming the configuration requests broker acks). When a response is received, the entry from the wait response queue is removed regardless of whether the response is a success or failure. Then, it checks whether it should retry, if retries are exhausted, it calls the registered client callback.

If the above is correct and we examine the proposed solution (`max.inflights=1`) then it seems to me that if the client code is producing messages fast enough then the broker send queue could have multiple buffers waiting, blocked by `max.inflights` vs. the size of the wait response queue. If an error is received and retries are exhausted, the blocking entry will be removed from the wait response queue allowing the next item on the broker send queue to be transmitted. If the error was transient (e.g. no leader), the send of the next buffer could succeed, violating the desired order preservation.

Am I missing something in the code that would prevent that second buffer from being sent if the first one fails?

Thanks!

**Comment by aaronhays** [comment]
@edenhill - when you get a chance, can you please verify whether my last comment holds water or if I'm missing something? This is still an outstanding concern for us and I'd like to be sure that the `max.inflights=1` trick will work. Thanks!

**Comment by edenhill** [comment]
@aaronhays Looking into the code to verify that you are correct, which I thought you were, shows that there are problems with the retry-path in correlation with max.inflight=1. The retry buffer may be placed after other messages, thus resulting in out-of-order despite max.inflight=1.
The fix requires a bigger change but must be fixed in the next release.
Created separate issue for this:
https://github.com/edenhill/librdkafka/issues/1092

**Comment by 2pl** [comment]
I'm currently trying to achieve guaranteed ordering as well. 
I understand the rationale behind setting max.in.flight=1: a transient error could lead to ProduceRequest 1 being rejected while ProduceRequest 2 on the same topic:partition succeeds. The theoretical risk exists. Now what are the actual broker failure scenario that could trigger it ? Would all of those scenario involve a change of leadership for the partition ? If they do involve a change of leadership, is that something that could be leveraged to offer guaranteed ordering with max.in.flight > 1 ?

**Comment by edenhill** [comment]
At the current state of things there is no absolute-no-reordering-guarantee in librdkafka.
The goal is to solve this properly with Idempotent producer and EOS (Exactly Once Semantics)

Here is one case (thanks @ijuma) which might cause reordering:
https://issues.apache.org/jira/browse/KAFKA-3197

**Comment by 2pl** [comment]
Thanks for your reply. I realize the best approach for librdkafka is to embrace Exactly Once. But it will take time before [Kafka Exactly-Once KIP-98](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging) hits the master branch.

So I guess what I'm trying to understand here is the magnitude of the risk, and how can I leverage existing APIs to detect and mitigate. So far I only found cases of reordering in the context of a change of partition leadership. Can someone think of a case where an out-of-order processing would occur with no change of partition leadership ?

**Comment by edenhill** [comment]
Idempotent Producer (#2001) solves message ordering guarantees
